WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainStateHolder.kt

```diff
@@ -16,161 +16,215 @@
-            AppAction.PlayMusicFinished,
-            AppAction.PlayRecordingFinished,
-            AppAction.RecordingFinished -> AppState.Ready(transitionInstantly = false)
-            AppAction.PermissionResultReturned ->
-                // Check if the user granted the permission
-                if (
-                    ContextCompat.checkSelfPermission(application, Manifest.permission.RECORD_AUDIO) ==
-                    PackageManager.PERMISSION_GRANTED
-                ) {
-                    // The user granted the permission, continue on to start recording
-                    AppState.Recording
-                } else {
-                    // We have confirmation now that the user denied the permission
-                    isPermissionDeniedMutableStateFlow.value = true
-                    AppState.Ready(transitionInstantly = false)
+                AppAction.PermissionResultReturned ->
+                    // Check if the user granted the permission
+                    if (
+                        ContextCompat.checkSelfPermission(activity, Manifest.permission.RECORD_AUDIO) ==
+                        PackageManager.PERMISSION_GRANTED
+                    ) {
+                        // The user granted the permission, continue on to start recording
+                        recordUpdatingState()
+                    } else {
+                        // We have confirmation now that the user denied the permission
+                        isPermissionDenied = true
+                        appState = AppState.Ready(transitionInstantly = false)
+                    }
+            }
+        }
+    }
+
+    /**
+     * A helper function to record, updating the progress state while recording.
+     *
+     * This requires the [Manifest.permission.RECORD_AUDIO] permission to run.
+     */
+    @RequiresPermission(Manifest.permission.RECORD_AUDIO)
+    private suspend fun recordUpdatingState() {
+        // We have the permission, we can start recording now
+        appState = AppState.Recording
+
+        coroutineScope {
+            // Kick off a parallel job to record
+            val recordingJob = launch { soundRecorder.record() }
+
+            val delayPerTickMs = MAX_RECORDING_DURATION.toMillis() / NUMBER_TICKS
+            val startTime = System.currentTimeMillis()
+
+            repeat(NUMBER_TICKS) { index ->
+                recordingProgress = index.toFloat() / NUMBER_TICKS
+                delay(startTime + delayPerTickMs * (index + 1) - System.currentTimeMillis())
+            }
+            // Update the progress to be complete
+            recordingProgress = 1f
+
+            // Stop recording
+            recordingJob.cancel()
+        }
+
+        appState = AppState.Ready(transitionInstantly = false)
+    }
+
+    /**
+     * Plays the MP3 file embedded in the application, updating the state to reflect the playing.
+     */
+    private suspend fun playMusicUpdatingState() {
```

Comment:
`playMusicUpdatingState` doesn't seem like a good method name. Maybe simply have `playMusic` and defer the updating state logic to the caller of the function?

Reply:
Ditto for `recordUpdatingState`

Reply:
Done in the latest commit!